### ISSUE TITLE
Extremely low loss for the encoder but no differentiating power found while training the classifier

### ISSUE THREAD

**Issue by hominpark** [sayakpaul/Supervised-Contrastive-Learning-in-TensorFlow-2#4]
First of all, thank you so much for the great implementation.

While I was working on my project exploiting your scripts, I found the following issues:

1) I have two different datasets, one is a dense data while the other is a sparse data
2) Encoder training loss is less than 2.0e-12 for both datasets
3) But the classifier accuracy is different. For the dense data it achieved >94% binary accuracy while the sparse data attained ~50% binary accuracy. Both went through more than 10,000 epochs.

I am very curious how this is possible since the loss is extremely small for both.

Any thoughts?

**Comment by sayakpaul** [comment]
That's indeed very strange to me. 

I would take a tSNE projection of the encoded representations of the sparse data to see if there's any regularity in the data. If you fail to see any regularity probably tinkering with the model architecture would be helpful. But that said, it still does not explain the low loss part (maybe it's overfitting to the validation set too?). Without taking a closer look, it's hard for me to suggest anything more concrete.

**Comment by hominpark** [comment]
Thank you for a quick reply.

I tried tSNE and indeed the different classes within the data is not differentiable (two classes are all mixed up). Yes, it is very strange to me as well. Hmm..

Have you ever tried to train the network using two classes that are very subtly different (e.g., golden retriever and labrador)?

**Comment by hominpark** [comment]
Plus, do you have to ensure the class balance within the batch? I figured using a large batch size leads to loss = nan, thus I reduced the batch size to 32.

**Comment by sayakpaul** [comment]
I see. That's something I didn't experiment with. 

> Plus, do you have to ensure the class balance within the batch?

I think the contrastive loss in phase 1 should be sufficient enough to figure that part out. Here are two things I would try - 
* Use `supervised_nt_xent_loss()` as the loss function. I would initially keep the temperature to 0.5. 
* Introduce data augmentation in phase 1 of the training.

**Comment by hominpark** [comment]
OK, thanks for the suggestions. I will give it a try and let you know the results :)

**Comment by hominpark** [comment]
> 
> 
> I see. That's something I didn't experiment with.
> 
> > Plus, do you have to ensure the class balance within the batch?
> 
> I think the contrastive loss in phase 1 should be sufficient enough to figure that part out. Here are two things I would try -
> 
>     * Use `supervised_nt_xent_loss()` as the loss function. I would initially keep the temperature to 0.5.
> 
>     * Introduce data augmentation in phase 1 of the training.

A quick update on the suggested methods.

1) Since my project is the suitable for data augmentation, I tried to use the supervised nt xent loss with varying temperatures including 0.5 and 0.1.
2) Unlike supervised contrastive loss, this new method converges to relatively higher loss values (e.g, 24 for temp = 0.5 and 3 for temp = 0.1).
3) classification accuracy is around 50% (no better than random guess) after 10,000 epochs.

While the problem I am tackling is a difficult one due to the very subtle difference between two classes, I am still confused how it was possible to attain such a low loss with the supervised contrastive loss yet have no differentiating power. Maybe, as you mentioned, it was extensively over-fitted to the training set?

**Comment by sayakpaul** [comment]
Yes, I think so. 

Did you use any augmentation transforms in the phase-1 training? From the information you provided it seems like the network might need a bit of regularization. I would start adding weight decay as a regularizer to see how it would affect the performance. 

Additionally, a trivial mistake I often make is I forget to scale the pixels appropriately. I am guessing you did ensure that.
